torch/csrc/profiler/python/init.cpp

```diff
@@ -112,6 +112,107 @@ struct type_caster<std::shared_ptr<torch::CapturedTraceback>> {
 namespace torch {
 namespace profiler {
 
+/* [NOTE: RecordFunctionFast]
+ * This is an alternate way to call record_function from python.
+ * The torch.profiler.record_function context manager is slow (~14us on
+ * benchmarks in Aug 2023), which is usually fine for module-level annotations
+ * in python, but slow for per-op annotations. Part of the reason it is slow is
+ * because the calls go through the dispatcher, in order to make the
+ * record_function calls work with torchscript.
+ *
+ * This implementation doesn't go through the dispatcher and so it won't work
+ * with any feature relying on the dispatcher (e.g. torchscript or
+ * torch.compile)
+ *
+ * An alternate solution would be to implement a python context manager that
+ * calls into C++ for the enter/exit function:
+ *    @contextlib.contextmanager
+ *    def record_function_fast(name):
+ *      rf = torch._C._record_function_fast_enter(name)
+ *      try:
+ *        yield
+ *      finally:
+ *        torch._C._record_function_fast_exit(rf)
+ * The C++ implementation here is faster by ~0.2-0.4us per context manager.
+ */
+
+namespace {
+struct RecordFunctionFast {
+  PyObject_HEAD PyObject* name;
+  std::unique_ptr<at::RecordFunction> guard;
+};
+
+PyObject* RecordFunctionFast_new(
+    PyTypeObject* subtype,
+    PyObject* args,
+    PyObject* kwargs) {
+  RecordFunctionFast* self = (RecordFunctionFast*)subtype->tp_alloc(subtype, 0);
+  if (self != nullptr) {
+    self->name = nullptr;
+    self->guard.reset();
+  }
+  return (PyObject*)self;
+}
+
+int RecordFunctionFast_init(
+    PyObject* selfGeneric,
+    PyObject* args,
+    PyObject* kwargs) {
+  auto self = (RecordFunctionFast*)selfGeneric;
+  constexpr const char* kwlist[] = {"name", nullptr};
+  PyObject* name = nullptr;
+  if (!PyArg_ParseTupleAndKeywords(
+          args, kwargs, "O", const_cast<char**>(kwlist), &name)) {
+    return -1;
+  }
+  if (name) {
+    Py_INCREF(name);
+    THPUtils_checkString(name);
```

Comment:
```suggestion
    TORCH_CHECK(THPUtils_checkString(name), "The name passed to RecordFunction must be a string");
    Py_INCREF(name);
```

The check is never raising and error and we should only incref if no error is raised.